// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: rill/admin/v1/ai.proto

package adminv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on CompleteRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CompleteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CompleteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CompleteRequestMultiError, or nil if none found.
func (m *CompleteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CompleteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetMessages() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, CompleteRequestValidationError{
						field:  fmt.Sprintf("Messages[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, CompleteRequestValidationError{
						field:  fmt.Sprintf("Messages[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return CompleteRequestValidationError{
					field:  fmt.Sprintf("Messages[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	for idx, item := range m.GetTools() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, CompleteRequestValidationError{
						field:  fmt.Sprintf("Tools[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, CompleteRequestValidationError{
						field:  fmt.Sprintf("Tools[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return CompleteRequestValidationError{
					field:  fmt.Sprintf("Tools[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return CompleteRequestMultiError(errors)
	}

	return nil
}

// CompleteRequestMultiError is an error wrapping multiple validation errors
// returned by CompleteRequest.ValidateAll() if the designated constraints
// aren't met.
type CompleteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CompleteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CompleteRequestMultiError) AllErrors() []error { return m }

// CompleteRequestValidationError is the validation error returned by
// CompleteRequest.Validate if the designated constraints aren't met.
type CompleteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CompleteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CompleteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CompleteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CompleteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CompleteRequestValidationError) ErrorName() string { return "CompleteRequestValidationError" }

// Error satisfies the builtin error interface
func (e CompleteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCompleteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CompleteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CompleteRequestValidationError{}

// Validate checks the field values on CompleteResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *CompleteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CompleteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CompleteResponseMultiError, or nil if none found.
func (m *CompleteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CompleteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetMessage()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CompleteResponseValidationError{
					field:  "Message",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CompleteResponseValidationError{
					field:  "Message",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMessage()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CompleteResponseValidationError{
				field:  "Message",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CompleteResponseMultiError(errors)
	}

	return nil
}

// CompleteResponseMultiError is an error wrapping multiple validation errors
// returned by CompleteResponse.ValidateAll() if the designated constraints
// aren't met.
type CompleteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CompleteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CompleteResponseMultiError) AllErrors() []error { return m }

// CompleteResponseValidationError is the validation error returned by
// CompleteResponse.Validate if the designated constraints aren't met.
type CompleteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CompleteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CompleteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CompleteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CompleteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CompleteResponseValidationError) ErrorName() string { return "CompleteResponseValidationError" }

// Error satisfies the builtin error interface
func (e CompleteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCompleteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CompleteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CompleteResponseValidationError{}
